MatrixSDK/Background/Store/SyncResponseFileStore/SyncResponseFileStore.swift

```diff
@@ -0,0 +1,167 @@
+        guard let jsonString = fileContents else {
+            return nil
+        }
+        guard let json = MXTools.deserialiseJSONString(jsonString) as? [AnyHashable: Any] else {
+            return nil
+        }
+        return MXSyncResponse(fromJSON: json)
+    }
+    
+    private func saveSyncResponse(_ syncResponse: MXSyncResponse?) {
+        guard let filePath = filePath else {
+            return
+        }
+        
+        guard let syncResponse = syncResponse else {
+            try? FileManager.default.removeItem(at: filePath)
+            return
+        }
+        SyncResponseFileStoreConstants.fileOperationQueue.async {
+            try? syncResponse.jsonString()?.write(to: self.filePath,
+                                                  atomically: true,
+                                                  encoding: SyncResponseFileStoreConstants.fileEncoding)
+        }
+    }
+    
+}
+
+//  MARK: - SyncResponseStore
+
+extension SyncResponseFileStore: SyncResponseStore {
+    
+    public func open(withCredentials credentials: MXCredentials) {
+        self.credentials = credentials
+        self.setupFilePath()
+    }
+    
+    public var syncResponse: MXSyncResponse? {
+        return readSyncResponse()
+    }
+    
+    public func event(withEventId eventId: String, inRoom roomId: String) -> MXEvent? {
+        guard let response = syncResponse else {
+            return nil
+        }
+        
+        var allEvents: [MXEvent] = []
+        if let joinedRoomSync = response.rooms.join[roomId] {
+            allEvents.appendIfNotNil(contentsOf: joinedRoomSync.state?.events)
+            allEvents.appendIfNotNil(contentsOf: joinedRoomSync.timeline?.events)
+            allEvents.appendIfNotNil(contentsOf: joinedRoomSync.accountData?.events)
+        }
+        if let invitedRoomSync = response.rooms.invite[roomId] {
+            allEvents.appendIfNotNil(contentsOf: invitedRoomSync.inviteState?.events)
+        }
+        if let leftRoomSync = response.rooms.leave[roomId] {
+            allEvents.appendIfNotNil(contentsOf: leftRoomSync.state?.events)
+            allEvents.appendIfNotNil(contentsOf: leftRoomSync.timeline?.events)
+            allEvents.appendIfNotNil(contentsOf: leftRoomSync.accountData?.events)
+        }
+        
+        let result = allEvents.first(where: { eventId == $0.eventId })
+        result?.roomId = roomId
+        return result
+    }
+    
+    public func update(with response: MXSyncResponse?) {
+        guard filePath != nil else {
+            return
+        }
+        
+        guard let response = response else {
+            //  Return if no new response
+            return
+        }
+        if let syncResponse = syncResponse {
+            //  current sync response exists, merge it with the new response
+            var dictionary = NSDictionary(dictionary: syncResponse.jsonDictionary())
+            dictionary = dictionary + NSDictionary(dictionary: response.jsonDictionary())
```

Comment:
There is more logic to apply.
1/ if the there is gap (limited = true) in a timeline, we need to delete messages already in SyncResponseFileStore. This ensures message continuity in the timeline.

Reply:
It'll be handled by `MXEventTimeline` when processing the sync response. I think we should keep this as is and rely on that.